### ISSUE TITLE
Ratatui's Vision

### ISSUE THREAD

**Issue by orhun** [ratatui-org/ratatui#1321]
## Problem

Ratatui should have a clear vision regarding how we approach problems and incorporate new features. I wanted to create this issue to share common ideas among maintainers and to clarify things for new contributors who would like to learn more about how things work. This will also help with maintenance, as it will serve as a guide for prioritization and hopefully get everyone on the same page. If you think this issue is missing any points or if you disagree with anything, feel free to discuss it below.

## Solution

Like any other open-source project, we should have guiding principles that we follow when determining our next steps.

I would like to draw your attention to the two most important areas we should be focusing on as a TUI library.

### 1. Ease of Use

There is a fine line between allowing customizations and making something difficult enough to create a frustrating development experience. Right now, I feel like we are standing in the middle, which can be further improved by addressing the following usability-related issues in the long term:

- #1287
- #174
- #293
- #1289 (abstraction over the backends + reducing the boilerplate)
- Simplify the example code, make them consistent to make it easier to read / write the relevant bits.

These and other _universally_ useful issues should be prioritized in development.

### 2. Visual Appeal

_What's a TUI library without a shiny UI?_

We want to draw more attention to Ratatui by making it prettier, and I believe this can be achieved by:

- Making the examples look more polished.
- Allowing widget customization, such as in #1168 and #605.
- Drawing inspiration from other TUI libraries (e.g., Textual, Bubbletea).

### Performance improvements

The current state of the library is not focused around zero allocations or generating the fastest LLVM IR. We take the performance into account while adding new features as long as that are aligned with an actual problem in a real application.

We are willing to make meaningful performance improvements if benchmarks show they are valuable. However, for smaller performance improvements, we will evaluate them on a case-by-case basis, considering factors such as maintainability and potential breakage.

### Breaking changes

We may need to consider breaking changes as we explore these improvements, as they could add significant value in the long term. However, making changing an existing API that would impact large amounts of users is not actionable unless it is really useful. 

### What are the next steps?

We should break down these larger goals into smaller, actionable tasks in a roadmap, ideally with some milestones as well, so we know what to do and when to do it next.

**Comment by nickspurry** [comment]
A point I think is missing from the above discussion is input widgets. At the moment, all widgets are focussed on display of information. The other obvious use of a TUI is to capture information. The user_input example demonstrates this by manually manipulating the display of a widget. While it works, the boilerplate code is cumbersome for a task that is probably fairly common. There are other common input types like drop downs, check boxes, radios etc. for which there are no examples. In my opinion, the project should decide whether development of input widgets is or isn't a project goal.

**Comment by joshka** [comment]
> A point I think is missing from the above discussion is input widgets

Definitely agree - I initially put some ideas that I thought would be explored outside of Ratatui in https://github.com/joshka/ratatui-widgets/. The naming of this as ratatui-widgets rather than tui-widgets was done with the intention that this might someday be a Ratatui core project rather than one that sits adjacent to Ratatui. There's also https://crates.io/crates/rat-salsa by @thscharler which is doing things a bit more rapidly in a similar area.

It would be definitely be reasonable to think of how Ratatui is positioned holistically and look to extract reasonably common approaches to user interaction (events, etc.), and to then make the long term approach to widgets (or whatever we end up calling them) include how they react to user input. This might include things like keyboard, mouse, focus, hit testing, ...

The two problems I see that we have by not being opinionated about this are:
- apps have to make their own opinions about how to do this. This causes friction for learning Ratatui and writing apps
- widget libraries need to create interaction code. This causes there to be multiple approaches which are duplicative / incompatible / confusing.

**Comment by orhun** [comment]
Agreed, incorporated those points above. Also, @ratatui/maintainers feel free to update the issue with relevant points 🙂

**Comment by thscharler** [comment]
I think the widgets can stay outside of ratatui proper, but some baselines are really missing. 

A good part of my work so far was coming up with answers to those questions. 

Specifically:
- support for displaying the cursor: [rat-cursor](https://docs.rs/rat-cursor)
- event handling: [rat-event](https://docs.rs/rat-event)
- focus handling: [rat-focus](https://docs.rs/rat-focus)

I think these are the main points, that would help interop of independently developed widgets. 

The application event loop and facilities of an application framework are not so important. I like my approach of just polling and distributing the events by just function calls, but I think not everyone will agree with me on that point.

**Comment by thscharler** [comment]
One more point, but not as important is theming. 
For this I think too that there is no one size fits all.

But what I added for all my widgets is a separate XXXStyle struct that holds everything that can be considered a 'style'
and can be set with one widget.styles() call. Since then I still have all the separate methods but hardly ever use them.

**Comment by thscharler** [comment]
If you are considering profound breaking changes here too I want to add:

* Make Rect and alike use __i32__ for the indexes. This would make it possible to partially scroll widgets left/top off screen. 

This can be emulated by creating a temporary Buffer and shifting it during copying. Works reasonably well for Widgets, but becomes unfeasable for StatefulWidgets that want to store areas for later interactions. Without very detailed knowledge of the internals of each such widget its almost impossible to track what the widget will do with the area given to render. As a result matching up a mouse position with the widget becomes hard to impossible. 

* If i32 is to big, i16 might do as well.

**Comment by joshka** [comment]
The problem with that change is that it's basically a break the world (every app, every widget) type change if done in a breaking way. At least I don't think there's a way to do that in a way that makes sense.

**Comment by kdheepak** [comment]
One idea that might be worth exploring is introducing an abstraction in between the buffer and the widgets. 

Currently widgets draw to the current buffer directly, but instead imagine widgets drew to a “layer”. With just one layer, widget render order will determine the outcome of the view. With multiple layers, you can draw items on top of each other by changing the `z` index of the layer. 

Additionally, layers could contain “transformations”. A transformations is a function that takes an x,y coordinate and returns another x,y coordinate. This would allow for drawing a layer that is rotated or flipped. It would allow for layers to be slid off screen. Or we could have a layer that always contains a -1,-1 offset from the layer on top of it, resulting in a shadow or 3-d effect. 

This idea is based on how notcurses works.

**Comment by thscharler** [comment]
> The problem with that change is that it's basically a break the world (every app, every widget) type change if done in a breaking way. At least I don't think there's a way to do that in a way that makes sense.

Yes, of course that's completely infeasable. Sorry for the bother. After some sleep, I think it wouldn't solve that problem anyway.

**Comment by orhun** [comment]
> Currently widgets draw to the current buffer directly, but instead imagine widgets drew to a “layer”. With just one layer, widget render order will determine the outcome of the view, the same way it is now with rendering to the buffer. But with multiple layers, you can draw items on top of each other by changing the z index of the layer.

This would be nice. I think it's also related to "making the Buffer a trait" idea from this [forum post](https://forum.ratatui.rs/t/brainstorming-redesigns/107).

> Additionally, layers could contain “transformations”. A transformations is a function that takes an x,y coordinate and returns another x,y coordinate. This would allow for drawing a layer that is rotated or flipped. It would allow for layers to be drawn partially off screen or magnified. Or we could have a layer that always contains a -1,-1 offset from the layer on top of it, resulting in a shadow or 3-d effect.

Sounds like a native alternative to [this](https://github.com/ratatui/ratatui-website/pull/725#issuecomment-2379751322) custom animation.

**Comment by joshka** [comment]
> Yes, of course that's completely infeasable. Sorry for the bother. After some sleep, I think it wouldn't solve that problem anyway.

No worries :) The underlying idea is good, it's just the breaking part that's a problem. At some point in the future we may consider working out a good way to correct a bunch of things like this.

**Comment by orhun** [comment]
As the lead maintainer of @ratatui, I would like to clarify my personal interaction with other projects as a part of this issue.
Recently I have been experiencing more situations where this has become confusing, so I think clearing the air once and for all would be helpful 1) for anyone who might have questions 2) for me mentally.

Since [#tui-rs#654](https://github.com/fdehau/tui-rs/issues/654) I have been doing a lot of social activities (posts, talks, podcasts, workshops) to spread the word of rats and reach to more people who are interested in doing terminal applications with Rust. I believe I have played a big part in growing the current ecosystem, and I'm genuinely pretty happy about the community we've built. I will keep doing these activities as much as I can.

While this outreach has had a positive impact, there was cases where I got mistaken for personal/political/controversial topics. 100% of the time I or @ratatui had nothing to do with that. I'm only focused on sharing cool tech around and I only care about working on the tech itself. In other words, if I interact with project X as a maintainer, that's only because the project has something to do with Ratatui, Rust or terminal related topics. There is no other underlying motive, and there never will be.

In other situations such as sponsored work (such as [JetBrains livestream](https://www.youtube.com/watch?v=UXnxzWbngz4), I will always make that context clear upfront.

I'm aware of the serious tone but I'm feeling a bit of pressure with my interactions recently and felt it was important to be transparent.
Hopefully this helps clarify my stance so we can move forward and keep the cheese rolling. 🧀
